# Методы отбора признаков в задачах Uplift-моделирования, основанные на важности признаков

В данном ноутбуке содержится исследование методов отбора признаков, основанных на важности признаков:
1) Отбор на основе shap-значений (shap > 0)
2) Отбор на основе permutation-importance (PI > 0)
3) Отбор на основе комбинации shap > 0 и PI > 0
4) Отбор с помощью feature importance из UpliftTree.

Во всех случаях рассчитываются значения важности признаков, после чего они подаются в T-Learner, базовой моделью которого является CatBoostClassifier

In [1]:
!pip install scikit-uplift

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.1/42.1 kB 1.3 MB/s eta 0:00:00


In [2]:
!pip install causalml

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 88.9/88.9 kB 2.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.9/8.9 MB 16.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 76.9/76.9 kB 1.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 15.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 55.2/55.2 kB 2.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.8/56.8 kB 1.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 269.8/269.8 kB 8.9 MB/s eta 0:00:00


In [3]:
!pip install catboost

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 97.1/97.1 MB 7.7 MB/s eta 0:00:00


In [4]:
from sklift.datasets import fetch_lenta
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.inspection import permutation_importance

from catboost import CatBoostClassifier, Pool
import shap

from causalml.metrics import plot_gain, auuc_score
from causalml.inference.tree import UpliftTreeClassifier, UpliftRandomForestClassifier
from causalml.inference.meta import BaseDRClassifier, BaseRClassifier, BaseSClassifier, BaseTClassifier, BaseXClassifier
from causalml.inference.iv import BaseDRIVRegressor

import warnings
warnings.filterwarnings("ignore")

ERROR:duecredit:Failed to import duecredit due to No module named 'duecredit'


In [5]:
SEED = 42

In [6]:
BASE_MODEL = CatBoostClassifier(
    loss_function='Logloss',
    verbose=False,
    random_state=SEED
  )

In [7]:
dataset = fetch_lenta()
data, target, treatment = dataset.data, dataset.target, dataset.treatment

Lenta dataset:   0%|          | 0.00/145M [00:00<?, ?iB/s]

### Загрузка и предобработка данных

После проведения EDA пришли к следующим выводам:

1) В датасете присутсвует дисбаланс классов: число записей, поповших в тестовую выбоку (воздействие было) составляет 75%, а контрольная группа составляет 25%.

2) Так же в датасете есть дисбаланс таргета: всего лишь 10% покупателей совершили целевое дейсвтие, в то время как другие 90% этого дейсвтия не совершали.

3) В датасете есть пропуски - их нужно заполнить. Так как минимальное значение в датасете - 0, заполним пропсуски значением -100.

4) В датасете есть только один категориальный признак - пол клиента, в данном признаке есть пропуски. Для данного признака присвоим значение 1, если пол мужской, 0, если женский. Пропущенные значения заполним -1.

5) В датасете есть 30 полных дубликатов, от них необходимо избавиться.

6) В результате превичного анализа датаффрейма никакие признаки не были удалены.

Предобработаем данные с учетом этих выводов:

In [8]:
# Преобразование признака gender
data['gender'] = data['gender'].map({'Male': 1, 'Female': 0})
data['gender'] = data['gender'].fillna(-1)

In [9]:
# Заполнение пропусков
data = data.fillna(-100)

In [10]:
data.isna().sum().sum()

np.int64(0)

In [11]:
# Удаление явных дубликатов
data = data.drop_duplicates()

In [12]:
# Разбиение на train, valid, test выборки

data['treatment'] = treatment
data['treatment'] = (data['treatment'] == 'test').astype(int)
data['target'] = target

train_val_idx, test_idx = train_test_split(data.index, test_size=0.2, random_state=SEED, stratify=data[['treatment', 'target']])

train_idx, val_idx = train_test_split(train_val_idx, test_size=0.25, random_state=SEED, stratify=data.loc[train_val_idx, ['treatment', 'target']])

X_train = data.drop(columns=['treatment', 'target']).loc[train_idx]
X_val = data.drop(columns=['treatment', 'target']).loc[val_idx]
X_test = data.drop(columns=['treatment', 'target']).loc[test_idx]

treatment_train = data['treatment'][train_idx]
treatment_val = data['treatment'][val_idx]
treatment_test = data['treatment'][test_idx]

y_train = target[train_idx]
y_val = target[val_idx]
y_test = target[test_idx]

### Отбор признаков

Теперь получим важность признаков с помощью встроенного в UpliftTree метода feature_importances_:

In [13]:
def get_importance_value_wooden_classifier(classifier, X_train, treatment_train, y_train):
  treatment_train_str = treatment_train.astype(str)
  treatment_test_str = treatment_test.astype(str)

  classifier.fit(X_train.values, treatment_train_str.values, y_train.values)

  feature_importances = pd.Series(classifier.feature_importances_, index=X_train.columns)
  selected_features = feature_importances[feature_importances > 0].index.tolist()

  return classifier, selected_features


In [14]:
def get_uplift(model_t, model_c, X, treatment, y,selected_features):

  X_selected = X[selected_features]

  pred_c = model_c.predict_proba(X_selected)[:, 1]
  pred_t = model_t.predict_proba(X_selected)[:, 1]

  uplift_pred = pred_t - pred_c

  pred_data = pd.DataFrame({'uplift': uplift_pred,
                            'y': y,
                            'treatment': treatment})

  auuc = auuc_score(pred_data, outcome_col='y', treatment_col='treatment')

  print('AUUC на train выборке:', auuc.squeeze())

  return uplift_pred, auuc

In [15]:
def train_tlearner(base_model, X_train, treatment_train, y_train, selected_features):
  X_c = X_train[treatment_train == 0][selected_features]
  y_c = y_train[treatment_train == 0]

  X_t = X_train[treatment_train == 1][selected_features]
  y_t = y_train[treatment_train == 1]

  model_c = base_model.copy()
  model_c.fit(X_c, y_c)

  model_t = base_model.copy()
  model_t.fit(X_t, y_t)

  uplift_pred_train, auuc_train = get_uplift(model_t, model_c, X_train, treatment_train, y_train, selected_features)


  return model_t, model_c, uplift_pred_train, auuc_train

Проведем эксперимент с UpliftTreeClassifier

In [16]:
uplift_tree = UpliftTreeClassifier(
      max_depth=5,
      min_samples_leaf=100,
      min_samples_treatment=50,
      n_reg=100,
      evaluationFunction='KL',
      control_name='0',
      random_state=SEED
  )

In [17]:
uplift_tree, selected_features_by_uplift_tree = get_importance_value_wooden_classifier(uplift_tree, X_train, treatment_train, y_train)

print(f'Число отобранных признаков {len(selected_features_by_uplift_tree)}')

Число отобранных признаков 14


In [18]:
selected_features_by_uplift_tree

['cheque_count_12m_g33',
 'cheque_count_12m_g45',
 'cheque_count_3m_g20',
 'crazy_purchases_cheque_count_3m',
 'k_var_count_per_cheque_3m_g24',
 'k_var_count_per_cheque_6m_g27',
 'k_var_days_between_visits_15d',
 'k_var_disc_share_3m_g44',
 'k_var_sku_price_15d_g34',
 'k_var_sku_price_1m_g34',
 'k_var_sku_price_3m_g27',
 'response_sms',
 'response_viber',
 'sale_sum_3m_g24']

In [19]:
model_t_tree, model_c_tree, uplift_pred_train_tree, auuc_train_tree = \
          train_tlearner(BASE_MODEL, X_train, treatment_train, y_train, selected_features_by_uplift_tree)

AUUC на train выборке: 5.872463483763391


In [20]:
uplift_pred_test_tree, auuc_test_tree = get_uplift(model_t_tree, model_c_tree, X_test, treatment_test, y_test, selected_features_by_uplift_tree)

AUUC на train выборке: 0.7133049312066355


In [21]:
auuc_test_tree

,0
uplift,0.713305


Попробуем посмотреть важность для UpliftForestClassifier

In [22]:
uplift_forest = UpliftRandomForestClassifier(
    max_depth=5,
    min_samples_leaf=100,
    min_samples_treatment=50,
    n_reg=100,
    evaluationFunction='KL',
    control_name='0',
    random_state=SEED
)


In [23]:
uplift_forest, selected_features_by_uplift_forest = get_importance_value_wooden_classifier(uplift_forest, X_train, treatment_train, y_train)

print(f'Число отобранных признаков {len(selected_features_by_uplift_forest)}')

Число отобранных признаков 92


In [24]:
selected_features_by_uplift_forest

['age',
 'cheque_count_12m_g20',
 'cheque_count_12m_g21',
 'cheque_count_12m_g25',
 'cheque_count_12m_g33',
 'cheque_count_12m_g38',
 'cheque_count_12m_g42',
 'cheque_count_12m_g45',
 'cheque_count_12m_g46',
 'cheque_count_12m_g56',
 'cheque_count_3m_g25',
 'cheque_count_3m_g79',
 'cheque_count_6m_g20',
 'cheque_count_6m_g21',
 'cheque_count_6m_g25',
 'cheque_count_6m_g33',
 'cheque_count_6m_g38',
 'cheque_count_6m_g40',
 'cheque_count_6m_g42',
 'cheque_count_6m_g57',
 'cheque_count_6m_g58',
 'cheque_count_6m_g79',
 'crazy_purchases_cheque_count_12m',
 'crazy_purchases_cheque_count_6m',
 'crazy_purchases_goods_count_12m',
 'crazy_purchases_goods_count_6m',
 'disc_sum_6m_g34',
 'food_share_15d',
 'k_var_cheque_15d',
 'k_var_cheque_category_width_15d',
 'k_var_count_per_cheque_15d_g34',
 'k_var_count_per_cheque_1m_g44',
 'k_var_count_per_cheque_1m_g49',
 'k_var_count_per_cheque_3m_g24',
 'k_var_count_per_cheque_3m_g27',
 'k_var_count_per_cheque_3m_g41',
 'k_var_count_per_cheque_3m_g44',


In [25]:
model_t_forest, model_c_forest, uplift_pred_train_forest, auuc_train_forest \
                 = train_tlearner(BASE_MODEL, X_train, treatment_train, y_train, selected_features_by_uplift_forest)

AUUC на train выборке: 8.431128375119986


In [26]:
uplift_pred_test_forest, auuc_test_forest \
      = get_uplift(model_t_forest, model_c_forest, X_test, treatment_test, y_test, selected_features_by_uplift_forest)

AUUC на train выборке: 0.6440905052062644


In [27]:
auuc_test_forest

,0
uplift,0.644091


Теперь Meta-Learners

In [28]:
# BASE_MODEL = CatBoostClassifier(
#     loss_function='Logloss',
#     verbose=False,
#     random_state=SEED
# )

In [29]:
s_learner = BaseSClassifier(learner=BASE_MODEL.copy())
t_learner = BaseTClassifier(learner=BASE_MODEL.copy())
r_learner = BaseRClassifier(outcome_learner=BASE_MODEL.copy(), effect_learner=BASE_MODEL.copy())
x_learner = BaseXClassifier(outcome_learner=BASE_MODEL.copy(), effect_learner=BASE_MODEL.copy())
dr_learner = BaseDRClassifier(learner=BASE_MODEL.copy())
driv_learner = BaseDRIVRegressor(learner=BASE_MODEL.copy())

Вариант расчета важности - permutation, shapley, feature_importance

In [30]:
def permutation_importance_selection(base_learner, X_val, treatment_val, y_val, n_repeats=2):
  feature_names = X_val.columns.tolist()

  def scorer(model, X, y):
      uplift_pred = model.predict(np.array(X).copy())
      uplift_pred = np.asarray(uplift_pred).ravel()

      pred_data = pd.DataFrame({
          'uplift': uplift_pred,
          'y': np.asarray(y).ravel(),
          'treatment': np.asarray(treatment_val).ravel()
      })

      auuc = auuc_score(pred_data, outcome_col='y', treatment_col='treatment')
      return float(auuc.iloc[0] if hasattr(auuc, 'iloc') else auuc)

  perm_importance = permutation_importance(
      base_learner, X_val, y_val,
      scoring=scorer,
      n_repeats=n_repeats,
      random_state=42,
      n_jobs=-1
  )

  importance_df = pd.DataFrame({
      'feature': feature_names,
      'importance_mean': perm_importance.importances_mean,
      'importance_std': perm_importance.importances_std
  })


  importance_df = importance_df.sort_values('importance_mean', ascending=False).reset_index(drop=True)

  selected_features = importance_df[importance_df['importance_mean'] > 0]['feature'].values.tolist().copy()

  return selected_features

In [31]:
# def shap_selection(base_learner, X_val, model_name):
#   if model_name == 's_learner':
#     explainer = shap.TreeExplainer(base_learner.model)
#     shap_values = explainer.shap_values(X_val)

#   elif model_name == 't_learner':
#     all_shap = []
#     for m in base_learner.models.values():
#       explainer = shap.TreeExplainer(m)
#       shap_vals = explainer.shap_values(X_val)
#       if isinstance(shap_vals, list):
#         shap_vals = shap_vals[1]
#       all_shap.append(shap_vals)
#     shap_values = np.mean(all_shap, axis=0)

#   elif model_name == 'x_learner':
#     all_shap = []
#     for m in list(base_learner.models_mu.values()) + list(base_learner.models_tau.values()):
#       explainer = shap.TreeExplainer(m)
#       shap_vals = explainer.shap_values(X_val)
#       if isinstance(shap_vals, list):
#         shap_vals = shap_vals[1]
#       all_shap.append(shap_vals)
#     shap_values = np.mean(all_shap, axis=0)

#   elif model_name == 'r_learner':
#     explainer = shap.TreeExplainer(base_learner.model_tau)
#     shap_values = explainer.shap_values(X_val)

#   elif model_name == 'dr_learner':
#     all_shap = []
#     for m in list(base_learner.models_mu.values()) + [base_learner.model_tau]:
#       explainer = shap.TreeExplainer(m)
#       shap_vals = explainer.shap_values(X_val)
#       if isinstance(shap_vals, list):
#         shap_vals = shap_vals[1]
#       all_shap.append(shap_vals)
#     shap_values = np.mean(all_shap, axis=0)

#   if isinstance(shap_values, list):
#     shap_values = shap_values[1]

#   shap_importance = np.abs(shap_values).mean(axis=0)
#   shap_importance_df = pd.DataFrame({'feature': X_val.columns.tolist(),
#                                      'importance': shap_importance})

#   selected_features = shap_importance_df[shap_importance_df.importance > 0]['feature'].values.tolist()

#   return selected_features


In [32]:
def shap_selection(base_learner, X_val, model_name):
  """
  Расчет SHAP важности для разных типов мета-лернеров
  """

  if model_name == 's_learner':
    # Для S-learner нужно получить обученную модель
    print("S-learner: попытка получить обученную модель")

    # Проверяем разные возможные места хранения модели
    if hasattr(base_learner, 'model') and base_learner.model is not None:
      try:
        # Пробуем создать explainer
        explainer = shap.TreeExplainer(base_learner.model)

        # Для S-learner модель могла обучаться с treatment фичей
        # Проверяем размерность
        if X_val.shape[1] + 1 == base_learner.model.n_features_in_:
          print("S-learner: модель ожидает treatment фичу, добавляем")
          # Добавляем фиктивный treatment (0) для предсказаний
          X_with_treatment = np.column_stack([np.zeros(X_val.shape[0]), X_val])
          shap_values = explainer.shap_values(X_with_treatment)
          # Убираем treatment фичу из результатов
          shap_values = shap_values[:, 1:]  # пропускаем первый столбец (treatment)
        else:
          shap_values = explainer.shap_values(X_val)

      except Exception as e:
        print(f"S-learner: ошибка при создании SHAP: {e}")

        # Альтернативный подход - используем .models если есть
        if hasattr(base_learner, 'models') and base_learner.models:
          print("S-learner: используем модели из .models")
          all_shap = []
          for name, model in base_learner.models.items():
            if model is not None:
              try:
                explainer = shap.TreeExplainer(model)
                shap_vals = explainer.shap_values(X_val)
                if isinstance(shap_vals, list):
                  shap_vals = shap_vals[1]
                all_shap.append(shap_vals)
              except:
                continue
          if all_shap:
            shap_values = np.mean(all_shap, axis=0)
          else:
            return []
    else:
      print("S-learner: модель не найдена в .model")

      # Пробуем найти модели в других атрибутах
      if hasattr(base_learner, 'models') and base_learner.models:
        print("S-learner: используем модели из .models")
        all_shap = []
        for name, model in base_learner.models.items():
          if model is not None:
            try:
              explainer = shap.TreeExplainer(model)
              shap_vals = explainer.shap_values(X_val)
              if isinstance(shap_vals, list):
                shap_vals = shap_vals[1]
              all_shap.append(shap_vals)
            except:
              continue
        if all_shap:
          shap_values = np.mean(all_shap, axis=0)
        else:
          return []
      else:
        return []

  elif model_name == 't_learner':
    # Для T-learner используем модели из словаря .models
    all_shap = []
    if hasattr(base_learner, 'models'):
      for name, model in base_learner.models.items():
        if model is not None:
          try:
            explainer = shap.TreeExplainer(model)
            shap_vals = explainer.shap_values(X_val)
            if isinstance(shap_vals, list):
              shap_vals = shap_vals[1]
            all_shap.append(shap_vals)
          except:
            continue
    if not all_shap:
      print("T-learner: модели не найдены или не поддерживают SHAP")
      return []
    shap_values = np.mean(all_shap, axis=0)

  elif model_name == 'x_learner':
    # Для X-learner используем outcome модели (.models_mu) и tau модели (.models_tau)
    all_shap = []
    if hasattr(base_learner, 'models_mu'):
      for model in base_learner.models_mu.values():
        if model is not None:
          try:
            explainer = shap.TreeExplainer(model)
            shap_vals = explainer.shap_values(X_val)
            if isinstance(shap_vals, list):
              shap_vals = shap_vals[1]
            all_shap.append(shap_vals)
          except:
            continue

    if hasattr(base_learner, 'models_tau'):
      for model in base_learner.models_tau.values():
        if model is not None:
          try:
            explainer = shap.TreeExplainer(model)
            shap_vals = explainer.shap_values(X_val)
            if isinstance(shap_vals, list):
              shap_vals = shap_vals[1]
            all_shap.append(shap_vals)
          except:
            continue

    if not all_shap:
      print("X-learner: модели не найдены или не поддерживают SHAP")
      return []
    shap_values = np.mean(all_shap, axis=0)

  elif model_name == 'r_learner':
    # Для R-learner используем модель эффекта .model_tau
    if hasattr(base_learner, 'model_tau') and base_learner.model_tau is not None:
      try:
        explainer = shap.TreeExplainer(base_learner.model_tau)
        shap_values = explainer.shap_values(X_val)
      except:
        print("R-learner: не удалось создать SHAP explainer")
        return []
    else:
      print("R-learner: модель не найдена")
      return []

  elif model_name == 'dr_learner':
    # Для DR-learner используем outcome модели (.models_mu) и модель эффекта (.model_tau)
    all_shap = []
    if hasattr(base_learner, 'models_mu'):
      for model in base_learner.models_mu.values():
        if model is not None:
          try:
            explainer = shap.TreeExplainer(model)
            shap_vals = explainer.shap_values(X_val)
            if isinstance(shap_vals, list):
              shap_vals = shap_vals[1]
            all_shap.append(shap_vals)
          except:
            continue

    if hasattr(base_learner, 'model_tau') and base_learner.model_tau is not None:
      try:
        explainer = shap.TreeExplainer(base_learner.model_tau)
        shap_vals = explainer.shap_values(X_val)
        if isinstance(shap_vals, list):
          shap_vals = shap_vals[1]
        all_shap.append(shap_vals)
      except:
        pass

    if not all_shap:
      print("DR-learner: модели не найдены или не поддерживают SHAP")
      return []
    shap_values = np.mean(all_shap, axis=0)

  else:
    print(f"Неизвестный тип модели: {model_name}")
    return []

  # Обрабатываем SHAP значения
  if isinstance(shap_values, list):
    shap_values = shap_values[1]  # для бинарной классификации берем класс 1

  # Рассчитываем среднюю абсолютную важность
  shap_importance = np.abs(shap_values).mean(axis=0)

  # Убеждаемся, что размерность совпадает
  if len(shap_importance) != X_val.shape[1]:
    print(f"Предупреждение: размерность SHAP ({len(shap_importance)}) не совпадает с числом признаков ({X_val.shape[1]})")
    # Если SHAP вернул больше признаков (из-за treatment), берем только нужные
    if len(shap_importance) > X_val.shape[1]:
      shap_importance = shap_importance[-X_val.shape[1]:]

  # Создаем DataFrame с результатами
  shap_importance_df = pd.DataFrame({
      'feature': X_val.columns.tolist(),
      'importance': shap_importance
  }).sort_values('importance', ascending=False)

  # Отбираем признаки с важностью > 0
  selected_features = shap_importance_df[shap_importance_df.importance > 0]['feature'].values.tolist()

  print(f"SHAP: отобрано {len(selected_features)} признаков")
  print("Топ-5 признаков по SHAP:")
  print(shap_importance_df.head(5))

  return selected_features

In [33]:
def get_importance(base_learner, model_name, X_train, treatment_train, y_train, X_val, treatment_val, y_val, importance_type='shapley'):

  base_learner.fit(X_train, treatment_train, y_train)

  cate = base_learner.predict(np.array(X_val).copy())

  if importance_type == 'feature_importance':
    if hasattr(base_learner, 'get_importance'):
      print('Рассчитывается Feature Importance')
      original_feature_names = X_val.columns.tolist()

      feature_mapping = pd.DataFrame({
          'original_name': original_feature_names,
          'mapped_name': [f"Feature_{i:03d}" for i in range(len(original_feature_names))]
      })

      importance_result = base_learner.get_importance(X_val, cate)
      importance_series = importance_result[1] if isinstance(importance_result, dict) else importance_result

      importance_df = pd.DataFrame({
          'mapped_name': importance_series.index,
          'importance': importance_series.values
      }).merge(feature_mapping, on='mapped_name', how='left')

      importance_df = importance_df.sort_values('importance', ascending=False)
      selected_features = importance_df[importance_df.importance > 0]['original_name'].values.tolist()
    else:
      print('У данного Мета-алгоритма нет встроенного метода get_importance')

  elif importance_type == 'shap':
    print('Вычисляем shap-values')
    selected_features = shap_selection(base_learner, X_val, model_name)

  elif importance_type == 'permutation':
    print('Вычисляем permutation importance')
    selected_features = permutation_importance_selection(base_learner, X_val, treatment_val, y_val, n_repeats=2)

  return base_learner, selected_features


In [34]:
def experiment_meta_learner(base_learner,
                            model_name,
                            X_train,
                            treatment_train,
                            y_train,
                            X_val,
                            treatment_val,
                            y_val,
                            X_test,
                            treatment_test,
                            y_test,
                            importance_type):

  fitted_learner, selected_features \
        = get_importance(base_learner, model_name, X_train, treatment_train, y_train, X_val, treatment_val, y_val, importance_type=importance_type)

  model_t, model_c, uplift_pred_train, auuc_train \
        = train_tlearner(BASE_MODEL, X_train, treatment_train, y_train, selected_features)


  uplift_pred_test, auuc_test \
        = get_uplift(model_t, model_c, X_test, treatment_test, y_test, selected_features)

  print(f'AUUC test: {auuc_test.squeeze()}')
  print(f'Число отобранных признаков: {len(selected_features)}')

  return fitted_learner, model_t, model_c, selected_features, auuc_train, auuc_test


Эксперименты с S-Learner

In [35]:
s_learner_fi, model_t_s_learner_fi, model_c_s_learner_fi, selected_features_s_learner_fi, \
      auuc_train_s_learner_fi, auuc_test_s_learner_fi = experiment_meta_learner(s_learner,
                                                                                's_learner',
                                                                                X_train,
                                                                                treatment_train,
                                                                                y_train,
                                                                                X_val,
                                                                                treatment_val,
                                                                                y_val,
                                                                                X_test,
                                                                                treatment_test,
                                                                                y_test,
                                                                                'feature_importance')

Рассчитывается Feature Importance
[LightGBM] [Warning] There are no meaningful features which satisfy the provided configuration. Decreasing Dataset parameters min_data_in_bin or min_data_in_leaf and re-constructing Dataset might resolve this warning.
[LightGBM] [Info] Total Bins 0
[LightGBM] [Info] Number of data points in the train set: 2, number of used features: 0
[LightGBM] [Info] Start training from score 0.500000
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are 

In [36]:
s_learner_pi, model_t_s_learner_pi, model_c_s_learner_pi, selected_features_s_learner_pi, \
      auuc_train_s_learner_pi, auuc_test_s_learner_pi = experiment_meta_learner(s_learner,
                                                                                's_learner',
                                                                                X_train,
                                                                                treatment_train,
                                                                                y_train,
                                                                                X_val,
                                                                                treatment_val,
                                                                                y_val,
                                                                                X_test,
                                                                                treatment_test,
                                                                                y_test,
                                                                                'permutation')

Вычисляем permutation importance
AUUC на train выборке: 8.387181161179775
AUUC на train выборке: 0.5117413085970506
AUUC test: 0.5117413085970506
Число отобранных признаков: 92


In [37]:
# s_learner_shap, model_t_s_learner_shap, model_c_s_learner_shap, selected_features_s_learner_shap, \
#       auuc_train_s_learner_shap, auuc_test_s_learner_shap = experiment_meta_learner(s_learner,
#                                                                                     's_learner',
#                                                                                     X_train,
#                                                                                     treatment_train,
#                                                                                     y_train,
#                                                                                     X_val,
#                                                                                     treatment_val,
#                                                                                     y_val,
#                                                                                     X_test,
#                                                                                     treatment_test,
#                                                                                     y_test,
#                                                                                     'shap')

Эксперименты с T-Learner

In [38]:
t_learner_fi, model_t_t_learner_fi, model_c_t_learner_fi, selected_features_t_learner_fi, \
      auuc_train_t_learner_fi, auuc_test_t_learner_fi = experiment_meta_learner(t_learner,
                                                                                't_learner',
                                                                                X_train,
                                                                                treatment_train,
                                                                                y_train,
                                                                                X_val,
                                                                                treatment_val,
                                                                                y_val,
                                                                                X_test,
                                                                                treatment_test,
                                                                                y_test,
                                                                                'feature_importance')

Рассчитывается Feature Importance
[LightGBM] [Warning] There are no meaningful features which satisfy the provided configuration. Decreasing Dataset parameters min_data_in_bin or min_data_in_leaf and re-constructing Dataset might resolve this warning.
[LightGBM] [Info] Total Bins 0
[LightGBM] [Info] Number of data points in the train set: 2, number of used features: 0
[LightGBM] [Info] Start training from score 0.500000
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are 

In [39]:
t_learner_pi, model_t_t_learner_pi, model_c_t_learner_pi, selected_features_t_learner_pi, \
      auuc_train_t_learner_pi, auuc_test_t_learner_pi = experiment_meta_learner(t_learner,
                                                                                't_learner',
                                                                                X_train,
                                                                                treatment_train,
                                                                                y_train,
                                                                                X_val,
                                                                                treatment_val,
                                                                                y_val,
                                                                                X_test,
                                                                                treatment_test,
                                                                                y_test,
                                                                                'permutation')

Вычисляем permutation importance
AUUC на train выборке: 8.404025177697813
AUUC на train выборке: 0.5349706147445782
AUUC test: 0.5349706147445782
Число отобранных признаков: 100


In [40]:
# t_learner_shap, model_t_t_learner_shap, model_c_t_learner_shap, selected_features_s_learner_shap, \
#       auuc_train_t_learner_shap, auuc_test_t_learner_shap = experiment_meta_learner(t_learner,
#                                                                                     't_learner',
#                                                                                     X_train,
#                                                                                     treatment_train,
#                                                                                     y_train,
#                                                                                     X_val,
#                                                                                     treatment_val,
#                                                                                     y_val,
#                                                                                     X_test,
#                                                                                     treatment_test,
#                                                                                     y_test,
#                                                                                     'shap')

Эксперименты с R-Learner

In [41]:
# r_learner_fi, model_t_r_learner_fi, model_c_r_learner_fi, selected_features_r_learner_fi, \
#       auuc_train_r_learner_fi, auuc_test_r_learner_fi = experiment_meta_learner(r_learner,
#                                                                                 'r_learner',
#                                                                                 X_train,
#                                                                                 treatment_train,
#                                                                                 y_train,
#                                                                                 X_val,
#                                                                                 treatment_val,
#                                                                                 y_val,
#                                                                                 X_test,
#                                                                                 treatment_test,
#                                                                                 y_test,
#                                                                                 'feature_importance')

In [42]:
# r_learner_pi, model_t_r_learner_pi, model_c_r_learner_pi, selected_features_r_learner_pi, \
#       auuc_train_r_learner_pi, auuc_test_r_learner_pi = experiment_meta_learner(r_learner,
#                                                                                 'r_learner',
#                                                                                 X_train,
#                                                                                 treatment_train,
#                                                                                 y_train,
#                                                                                 X_val,
#                                                                                 treatment_val,
#                                                                                 y_val,
#                                                                                 X_test,
#                                                                                 treatment_test,
#                                                                                 y_test,
#                                                                                 'permutation')

In [43]:
# r_learner_shap, model_t_r_learner_shap, model_c_r_learner_shap, selected_features_r_learner_shap, \
#       auuc_train_r_learner_shap, auuc_test_r_learner_shap = experiment_meta_learner(r_learner,
#                                                                                     'r_learner',
#                                                                                     X_train,
#                                                                                     treatment_train,
#                                                                                     y_train,
#                                                                                     X_val,
#                                                                                     treatment_val,
#                                                                                     y_val,
#                                                                                     X_test,
#                                                                                     treatment_test,
#                                                                                     y_test,
#                                                                                     'shap')

Эксперименты с X-Learner

In [44]:
# x_learner_fi, model_t_x_learner_fi, model_c_x_learner_fi, selected_features_x_learner_fi, \
#       auuc_train_x_learner_fi, auuc_test_x_learner_fi = experiment_meta_learner(x_learner,
#                                                                                 'x_learner',
#                                                                                 X_train,
#                                                                                 treatment_train,
#                                                                                 y_train,
#                                                                                 X_val,
#                                                                                 treatment_val,
#                                                                                 y_val,
#                                                                                 X_test,
#                                                                                 treatment_test,
#                                                                                 y_test,
#                                                                                 'feature_importance')

CatBoostError: catboost/private/libs/target/target_converter.cpp:410: Target with classes must contain only 2 unique values for binary classification

In [ ]:
# x_learner_pi, model_t_x_learner_pi, model_c_x_learner_pi, selected_features_x_learner_pi, \
#       auuc_train_x_learner_pi, auuc_test_x_learner_pi = experiment_meta_learner(x_learner,
#                                                                                 'x_learner',
#                                                                                 X_train,
#                                                                                 treatment_train,
#                                                                                 y_train,
#                                                                                 X_val,
#                                                                                 treatment_val,
#                                                                                 y_val,
#                                                                                 X_test,
#                                                                                 treatment_test,
#                                                                                 y_test,
#                                                                                 'permutation')

In [ ]:
# x_learner_shap, model_t_x_learner_shap, model_c_x_learner_shap, selected_features_x_learner_shap, \
#       auuc_train_x_learner_shap, auuc_test_x_learner_shap = experiment_meta_learner(x_learner,
#                                                                                     'x_learner',
#                                                                                     X_train,
#                                                                                     treatment_train,
#                                                                                     y_train,
#                                                                                     X_val,
#                                                                                     treatment_val,
#                                                                                     y_val,
#                                                                                     X_test,
#                                                                                     treatment_test,
#                                                                                     y_test,
#                                                                                     'shap')

Эксперименты с DR-Learner

In [ ]:
# dr_learner_fi, model_t_dr_learner_fi, model_c_dr_learner_fi, selected_features_dr_learner_fi, \
#       auuc_train_dr_learner_fi, auuc_test_dr_learner_fi = experiment_meta_learner(dr_learner,
#                                                                                   'dr_learner',
#                                                                                   X_train,
#                                                                                   treatment_train,
#                                                                                   y_train,
#                                                                                   X_val,
#                                                                                   treatment_val,
#                                                                                   y_val,
#                                                                                   X_test,
#                                                                                   treatment_test,
#                                                                                   y_test,
#                                                                                   'feature_importance')

In [ ]:
# dr_learner_pi, model_t_dr_learner_pi, model_c_dr_learner_pi, selected_features_dr_learner_pi, \
#       auuc_train_dr_learner_pi, auuc_test_dr_learner_pi = experiment_meta_learner(dr_learner,
#                                                                                   'dr_learner',
#                                                                                   X_train,
#                                                                                   treatment_train,
#                                                                                   y_train,
#                                                                                   X_val,
#                                                                                   treatment_val,
#                                                                                   y_val,
#                                                                                   X_test,
#                                                                                   treatment_test,
#                                                                                   y_test,
#                                                                                   'permutation')

In [ ]:
# dr_learner_shap, model_t_dr_learner_shap, model_c_dr_learner_shap, selected_features_dr_learner_shap, \
#       auuc_train_dr_learner_shap, auuc_test_dr_learner_shap = experiment_meta_learner(dr_learner,
#                                                                                       'dr_learner',
#                                                                                       X_train,
#                                                                                       treatment_train,
#                                                                                       y_train,
#                                                                                       X_val,
#                                                                                       treatment_val,
#                                                                                       y_val,
#                                                                                       X_test,
#                                                                                       treatment_test,
#                                                                                       y_test,
#                                                                                       'shap')

Эксперименты с DRIV-Learner

In [ ]:
# driv_learner_fi, model_t_driv_learner_fi, model_c_driv_learner_fi, selected_features_driv_learner_fi, \
#       auuc_train_driv_learner_fi, auuc_test_driv_learner_fi = experiment_meta_learner(driv_learner,
#                                                                                       'dr_learner',
#                                                                                       X_train,
#                                                                                       treatment_train,
#                                                                                       y_train,
#                                                                                       X_val,
#                                                                                       treatment_val,
#                                                                                       y_val,
#                                                                                       X_test,
#                                                                                       treatment_test,
#                                                                                       y_test,
#                                                                                       'feature_importance')

In [ ]:
# driv_learner_pi, model_t_driv_learner_pi, model_c_driv_learner_pi, selected_features_driv_learner_pi, \
#       auuc_train_driv_learner_pi, auuc_test_driv_learner_pi = experiment_meta_learner(driv_learner,
#                                                                                       'dr_learner',
#                                                                                       X_train,
#                                                                                       treatment_train,
#                                                                                       y_train,
#                                                                                       X_val,
#                                                                                       treatment_val,
#                                                                                       y_val,
#                                                                                       X_test,
#                                                                                       treatment_test,
#                                                                                       y_test,
#                                                                                       'permutation')

In [ ]:
# driv_learner_shap, model_t_driv_learner_shap, model_c_driv_learner_shap, selected_features_driv_learner_shap, \
#       auuc_train_driv_learner_shap, auuc_test_driv_learner_shap = experiment_meta_learner(driv_learner,
#                                                                                           'dr_learner',
#                                                                                           X_train,
#                                                                                           treatment_train,
#                                                                                           y_train,
#                                                                                           X_val,
#                                                                                           treatment_val,
#                                                                                           y_val,
#                                                                                           X_test,
#                                                                                           treatment_test,
#                                                                                           y_test,
#                                                                                           'shap')

**Вывод:**

1. Для данной части исследования использовали следующие методы отбора признаков:


*   Shap importance - отдельно для treatment и control моделей были рассчитаны shap значения всех признаков. После чего были оставлены только те признаки, shap-важность которых была больше 0
*   Permutation importance - аналогично предыдущему методу, только в качестве оценки важности признаков использовался метод permutation importance

* Shap + permutation - использовалось пересечение двух указанных выше методов

* UpliftTree - построили UpliftTreeClassifier, получили из него важность признаков. Оставили только те признаки, важность которых > 0.

После чего построили T-Learner на основе CatBoost на отобранных признаках.


2. Качество представленных методов отбора признаков значительно выше фильтрационных методов (где максимаьное значение auuc составляело 0.38)

3.  Ниже представлена таблица с метрикой auuc для реализованных методов и число вошедших признаков:


|Метод отбора признаков| AUUC | Число признаков treatment | Число признаков control |
|------| ----| --- | --- |
|Shap| 0.493202| 92 | 97 |
|Permutation| 0.563487| 179 | 173 |
|Shap + Permutation| 0.446839 | 90 | 89|
|UpliftTree importance| 0.713305| 14 | 14 |

Для последнего метода метрика auuc значительно выше.